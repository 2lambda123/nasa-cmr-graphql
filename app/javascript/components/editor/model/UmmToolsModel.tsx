/* eslint-disable dot-notation */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { UiSchema } from '@rjsf/utils'
import toolInformationUiSchema from '../data/configuration/uiSchemas/tools/tool_information'
import relatedUrlsUiSchema from '../data/configuration/uiSchemas/tools/related_urls'
import toolsConfiguration from '../data/configuration/uiForms/toolsConfiguration'
import organizationUiSchema from '../data/configuration/uiSchemas/tools/organization'
import UmmModel from './UmmModel'
import descriptiveKeywordsUiSchema from '../data/configuration/uiSchemas/tools/descriptive_keywords'
import potentialActionUiSchema from '../data/configuration/uiSchemas/tools/potential_action'
import toolContactsUISchema from '../data/configuration/uiSchemas/tools/tool_contacts'
import compatibilityAndUsabilityUiSchema from '../data/configuration/uiSchemas/tools/compatibility_and_usability'
import CustomArrayFieldTemplate from '../components/templates/CustomArrayFieldTemplate'
import schema from '../data/schemas/umm_tools_schema'

export default class UmmToolsModel extends UmmModel {
  constructor() {
    super(schema)
  }
  get documentTypeForDisplay() {
    return 'Tool Record'
  }
  get documentType() {
    return 'tool_drafts'
  }
  get formSections() {
    return toolsConfiguration
  }
  getFormSchema(): { [key: string]: object } {
    // use this function to edit schema
    return super.getFormSchema()
  }
  get uiSchema() {
    const base: UiSchema = {
      'ui:submitButtonOptions': { norender: true, submitText: 'Save' },
      'ui:ArrayFieldTemplate': CustomArrayFieldTemplate
    }
    if (this.currentSection.displayName === 'Related URLs') {
      const uiSchema: any = relatedUrlsUiSchema
      const urlUiSchema: any = uiSchema.RelatedURLs.items
      urlUiSchema['ui:service'] = this.service

      // Format
      const formatUiSchema: any = uiSchema.RelatedURLs.items.Format
      formatUiSchema['ui:service'] = this.service
      formatUiSchema['ui:keyword_scheme'] = 'granule_data_format'
      formatUiSchema['ui:keyword_scheme_column_names'] = ['short_name']

      // Mime Type
      const mimeTypeUiSchema: any = uiSchema.RelatedURLs.items.MimeType
      mimeTypeUiSchema['ui:service'] = this.service
      mimeTypeUiSchema['ui:keyword_scheme'] = 'mime_type'
      mimeTypeUiSchema['ui:keyword_scheme_column_names'] = ['mime_type']
      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Tool Information') {
      const uiSchema: any = toolInformationUiSchema
      const urlUiSchema: any = uiSchema.URL
      urlUiSchema['ui:service'] = this.service
      urlUiSchema['ui:keyword_scheme'] = 'rucontenttype'
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Descriptive Keywords') {
      const uiSchema = descriptiveKeywordsUiSchema
      const toolKeywords: any = uiSchema.ToolKeywords
      toolKeywords['ui:service'] = this.service

      toolKeywords['ui:keyword_scheme'] = 'science_keywords'
      toolKeywords['ui:picker_title'] = 'TOOL KEYWORD'
      toolKeywords['ui:scheme_values'] = [
        'ToolCategory', 'ToolTopic', 'ToolTerm', 'ToolSpecificTerm'
      ]
      toolKeywords['ui:keyword_scheme_column_names'] = ['toolkeywords', 'category', 'topic', 'term', 'variable_level_1', 'variable_level_2', 'variable_level_3']
      toolKeywords['ui:filter'] = (path: string[]) => path[1] !== 'EARTH SCIENCE'

      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Tool Organizations') {
      const uiSchema: any = organizationUiSchema
      const organizations: any = uiSchema.Organizations.items
      organizations['ui:service'] = this.service
      organizations['ui:keyword_scheme'] = 'providers'
      organizations['ui:keyword_scheme_column_names'] = ['short_name', 'long_name', 'url']
      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Tool Contacts') {
      const uiSchema: any = toolContactsUISchema
      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Potential Action') {
      const uiSchema: any = potentialActionUiSchema
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Compatibility And Usability') {
      const uiSchema: any = compatibilityAndUsabilityUiSchema
      return { ...uiSchema, ...base }
    }
    return { ...base }
  }

  // In this case the old preview sends the same section name that the react component uses (just lowercase), so
  // nothing needs to be migrated. once the old MMT preview is replaced, this 'migrate" functionality can be removed.
  migratedSectionName(sectionName: string): string {
    return sectionName
  }
}
