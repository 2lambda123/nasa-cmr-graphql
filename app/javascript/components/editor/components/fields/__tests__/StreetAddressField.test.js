import React from 'react'
import {
  act,
  render, screen
} from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import userEvent from '@testing-library/user-event'
import UmmToolsModel from '../../../model/UmmToolsModel'
import MetadataEditor from '../../../MetadataEditor'
import StreetAddressesField from '../StreetAddresseField'

describe('StreetAddressFieldTest', () => {
  it('basic test', async () => {
    const model = new UmmToolsModel()
    const editor = new MetadataEditor(model)
    const mockedOnChange = jest.fn()
    const props = {
      schema: {
        maxLength: 80,
        description: 'The street address description'
      },
      registry: {
        formContext: { editor }
      },
      onChange: mockedOnChange,
      formData: ['my first line', 'my second line', 'my third line']
    }
    const { container } = render(
      <BrowserRouter>
        <StreetAddressesField {...props} />
      </BrowserRouter>
    )
    expect(screen.getByTestId('street-address-field__title')).toHaveTextContent('The street address description')
    expect(screen.getByTestId('custom-text-widget__address-line-1--text-input')).toHaveValue('my first line')
    expect(screen.getByTestId('custom-text-widget__address-line-2--text-input')).toHaveValue('my second line')
    expect(screen.getByTestId('custom-text-widget__address-line-3--text-input')).toHaveValue('my third line')
    expect(container).toMatchSnapshot()
  })

  it('basic test with empty form data array', async () => {
    const model = new UmmToolsModel()
    const editor = new MetadataEditor(model)
    const mockedOnChange = jest.fn()
    const props = {
      schema: {
        maxLength: 80,
        description: 'The street address description'
      },
      registry: {
        formContext: { editor }
      },
      onChange: mockedOnChange,
      formData: []
    }
    const { container } = render(
      <BrowserRouter>
        <StreetAddressesField {...props} />
      </BrowserRouter>
    )
    expect(screen.getByTestId('street-address-field__title')).toHaveTextContent('The street address description')
    expect(screen.getByTestId('custom-text-widget__address-line-1--text-input')).toHaveValue('')
    expect(screen.getByTestId('custom-text-widget__address-line-2--text-input')).toHaveValue('')
    expect(screen.getByTestId('custom-text-widget__address-line-3--text-input')).toHaveValue('')
    expect(container).toMatchSnapshot()
  })

  it('basic test with no form data array', async () => {
    const model = new UmmToolsModel()
    const editor = new MetadataEditor(model)
    const mockedOnChange = jest.fn()
    const props = {
      schema: {
        maxLength: 80,
        description: 'The street address description'
      },
      registry: {
        formContext: { editor }
      },
      onChange: mockedOnChange
    }
    const { container } = render(
      <BrowserRouter>
        <StreetAddressesField {...props} />
      </BrowserRouter>
    )
    expect(screen.getByTestId('street-address-field__title')).toHaveTextContent('The street address description')
    expect(screen.getByTestId('custom-text-widget__address-line-1--text-input')).toHaveValue('')
    expect(screen.getByTestId('custom-text-widget__address-line-2--text-input')).toHaveValue('')
    expect(screen.getByTestId('custom-text-widget__address-line-3--text-input')).toHaveValue('')
    expect(container).toMatchSnapshot()
  })

  it('bubbles up changes.', async () => {
    const model = new UmmToolsModel()
    const editor = new MetadataEditor(model)
    const mockedOnChange = jest.fn()
    const props = {
      schema: {
        maxLength: 80,
        description: 'The street address description'
      },
      registry: {
        formContext: { editor }
      },
      onChange: mockedOnChange,
      formData: ['my first line', 'my second line', 'my third line']
    }
    const { container } = render(
      <BrowserRouter>
        <StreetAddressesField {...props} />
      </BrowserRouter>
    )
    const inputElement = screen.getByTestId('custom-text-widget__address-line-1--text-input')

    act(() => {
      userEvent.clear(inputElement)
      userEvent.type(inputElement, 'Cloudy day')
    })
    expect(inputElement.value).toBe('Cloudy day')
    expect(mockedOnChange).toBeCalledTimes(11)
    expect(mockedOnChange).toBeCalledWith(['Cloudy day', 'my second line', 'my third line'])

    const inputElement2 = screen.getByTestId('custom-text-widget__address-line-2--text-input')
    act(() => {
      userEvent.clear(inputElement2)
    })

    const inputElement3 = screen.getByTestId('custom-text-widget__address-line-3--text-input')
    act(() => {
      userEvent.clear(inputElement3)
      userEvent.type(inputElement3, 'Sunny day')
    })
    expect(mockedOnChange).toBeCalledTimes(22)
    expect(mockedOnChange).toBeCalledWith(['Cloudy day', 'Sunny day'])

    expect(container).toMatchSnapshot()
  })
})
