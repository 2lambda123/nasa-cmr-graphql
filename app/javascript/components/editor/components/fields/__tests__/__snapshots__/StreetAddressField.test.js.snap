// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreetAddressFieldTest basic test 1`] = `
<div>
  <span
    class="header-description"
    data-testid="street-address-field__title"
  >
    The street address description
  </span>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-1--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 1
      </span>
      <span>
        13
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-1--text-input"
      id="undefined_1"
      maxlength="80"
      name="Address Line 1"
      type="text"
      value="my first line"
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-1"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-2--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 2
      </span>
      <span>
        14
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-2--text-input"
      id="undefined_2"
      maxlength="80"
      name="Address Line 2"
      type="text"
      value="my second line"
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-2"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-3--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 3
      </span>
      <span>
        13
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-3--text-input"
      id="undefined_3"
      maxlength="80"
      name="Address Line 3"
      type="text"
      value="my third line"
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-3"
    />
  </div>
</div>
`;

exports[`StreetAddressFieldTest basic test with empty form data array 1`] = `
<div>
  <span
    class="header-description"
    data-testid="street-address-field__title"
  >
    The street address description
  </span>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-1--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 1
      </span>
      <span>
        0
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-1--text-input"
      id="undefined_1"
      maxlength="80"
      name="Address Line 1"
      type="text"
      value=""
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-1"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-2--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 2
      </span>
      <span>
        0
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-2--text-input"
      id="undefined_2"
      maxlength="80"
      name="Address Line 2"
      type="text"
      value=""
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-2"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-3--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 3
      </span>
      <span>
        0
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-3--text-input"
      id="undefined_3"
      maxlength="80"
      name="Address Line 3"
      type="text"
      value=""
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-3"
    />
  </div>
</div>
`;

exports[`StreetAddressFieldTest basic test with no form data array 1`] = `
<div>
  <span
    class="header-description"
    data-testid="street-address-field__title"
  >
    The street address description
  </span>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-1--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 1
      </span>
      <span>
        0
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-1--text-input"
      id="undefined_1"
      maxlength="80"
      name="Address Line 1"
      type="text"
      value=""
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-1"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-2--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 2
      </span>
      <span>
        0
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-2--text-input"
      id="undefined_2"
      maxlength="80"
      name="Address Line 2"
      type="text"
      value=""
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-2"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-3--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 3
      </span>
      <span>
        0
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-3--text-input"
      id="undefined_3"
      maxlength="80"
      name="Address Line 3"
      type="text"
      value=""
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-3"
    />
  </div>
</div>
`;

exports[`StreetAddressFieldTest bubbles up changes. 1`] = `
<div>
  <span
    class="header-description"
    data-testid="street-address-field__title"
  >
    The street address description
  </span>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-1--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 1
      </span>
      <span>
        10
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-1--text-input"
      id="undefined_1"
      maxlength="80"
      name="Address Line 1"
      type="text"
      value="Cloudy day"
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-1"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-2--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 2
      </span>
      <span>
        0
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-2--text-input"
      id="undefined_2"
      maxlength="80"
      name="Address Line 2"
      type="text"
      value=""
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-2"
    />
  </div>
  <div
    class="address-line col-md-12"
  >
    <div
      class="widget-header"
      data-testid="custom-text-widget__address-line-3--text-header"
    >
      <span
        class="metadata-editor-field-label null"
      >
        Address Line 3
      </span>
      <span>
        9
        /
        80
      </span>
    </div>
    <input
      class="custom-text-widget-input"
      data-testid="custom-text-widget__address-line-3--text-input"
      id="undefined_3"
      maxlength="80"
      name="Address Line 3"
      type="text"
      value="Sunny day"
    />
    <span
      class="widget-description"
      data-testid="custom-text-widget--description-field__address-line-3"
    />
  </div>
</div>
`;
